exporters/otlp/otlplog/otlploggrpc/config.go

```diff
@@ -104,10 +107,14 @@ func newConfig(options []Option) config {
 	}
 
 	// Apply environment value and default value
+	var scheme string
 	c.endpoint = c.endpoint.Resolve(
-		getEnv[string](envEndpoint, convEndpoint),
+		loadEnvEndpoint(func(u *url.URL) {
+			scheme = u.Scheme
+		}),
 		fallback[string](defaultEndpoint),
 	)
+	c.insecure = insecureFromScheme(c.insecure, scheme)
 	c.insecure = c.insecure.Resolve(
 		getEnv[bool](envInsecure, convInsecure),
 	)
```

Comment:
```suggestion
	c.insecure = c.insecure.Resolve(
		loadInsecureFromEnvEndpoint(envEndpoint),
		getEnv[bool](envInsecure, convInsecure),
	)
```

Could we implement a method called `loadInsecureFromEnvEndpoint` to load endpoint env and determine whether to set the insecure flag?

And, we don't need these lines:

https://github.com/open-telemetry/opentelemetry-go/pull/5739/files#diff-fe76d167f1601d93a1ce79db97be5dda69f605c24b7ec543202aaf74418b9223R110-R117

 This would make the code look clearer.

Reply:
Yeah, it looks cleaner!
But there are a little bit more boilerplate code, is it ok? Or maybe you could suggest a better way?

```
diff --git a/exporters/otlp/otlplog/otlploggrpc/config.go b/exporters/otlp/otlplog/otlploggrpc/config.go
index e1d03311..a78ff7a4 100644
--- a/exporters/otlp/otlplog/otlploggrpc/config.go
+++ b/exporters/otlp/otlplog/otlploggrpc/config.go
@@ -107,15 +107,12 @@ func newConfig(options []Option) config {
        }

        // Apply environment value and default value
-       var scheme string
        c.endpoint = c.endpoint.Resolve(
-               loadEnvEndpoint(func(u *url.URL) {
-                       scheme = u.Scheme
-               }),
+               loadEnvEndpoint(envEndpoint),
                fallback[string](defaultEndpoint),
        )
-       c.insecure = insecureFromScheme(c.insecure, scheme)
        c.insecure = c.insecure.Resolve(
+               loadInsecureFromEnvEndpoint(envEndpoint),
                getEnv[bool](envInsecure, convInsecure),
        )
        c.tlsCfg = c.tlsCfg.Resolve(
@@ -524,7 +521,7 @@ func loadCertificates(certPath, keyPath string) ([]tls.Certificate, error) {

 // loadEnvEndpoint loads and returns endpoint from envEndpoint variables.
 // It parses env value with url.Parse.
-func loadEnvEndpoint(fn func(*url.URL)) resolver[string] {
+func loadEnvEndpoint(envEndpoint []string) resolver[string] {
        return func(s setting[string]) setting[string] {
                if s.Set {
                        // Passed, valid, options have precedence.
@@ -539,10 +536,7 @@ func loadEnvEndpoint(fn func(*url.URL)) resolver[string] {
                                        continue
                                }

-                               s.Value = u.Host
-                               s.Set = true
-                               fn(u)
-                               return s
+                               return newSetting(u.Host)
                        }
                }
                return s
@@ -562,6 +556,28 @@ func insecureFromScheme(prev setting[bool], scheme string) setting[bool] {
        return prev
 }

+func loadInsecureFromEnvEndpoint(envEndpoint []string) resolver[bool] {
+       return func(s setting[bool]) setting[bool] {
+               if s.Set {
+                       // Passed, valid, options have precedence.
+                       return s
+               }
+
+               for _, key := range envEndpoint {
+                       if vStr := os.Getenv(key); vStr != "" {
+                               u, err := url.Parse(vStr)
+                               if err != nil {
+                                       otel.Handle(fmt.Errorf("invalid %s value %s: %w", key, vStr, err))
+                                       continue
+                               }
+
+                               return insecureFromScheme(s, u.Scheme)
+                       }
+               }
+               return s
+       }
+}
+
 func compressorToCompression(compressor string) Compression {
        c, err := convCompression(compressor)
        if err != nil {
```

Reply:
We only need to add the `loadInsecureFromEnvEndpoint` and `insecureFromScheme` methods. The `loadEnvEndpoint` method is not needed.

Here is the diff:

```diff
diff --git a/exporters/otlp/otlplog/otlploggrpc/config.go b/exporters/otlp/otlplog/otlploggrpc/config.go
index e1d03311..1d109e70 100644
--- a/exporters/otlp/otlplog/otlploggrpc/config.go
+++ b/exporters/otlp/otlplog/otlploggrpc/config.go
@@ -107,15 +107,12 @@ func newConfig(options []Option) config {
 	}
 
 	// Apply environment value and default value
-	var scheme string
 	c.endpoint = c.endpoint.Resolve(
-		loadEnvEndpoint(func(u *url.URL) {
-			scheme = u.Scheme
-		}),
+		getEnv[string](envEndpoint, convEndpoint),
 		fallback[string](defaultEndpoint),
 	)
-	c.insecure = insecureFromScheme(c.insecure, scheme)
 	c.insecure = c.insecure.Resolve(
+		loadInsecureFromEnvEndpoint(envEndpoint),
 		getEnv[bool](envInsecure, convInsecure),
 	)
 	c.tlsCfg = c.tlsCfg.Resolve(
@@ -387,15 +384,47 @@ func convCompression(s string) (Compression, error) {
 	return NoCompression, fmt.Errorf("unknown compression: %s", s)
 }
 
-// convInsecure converts s from string to bool without case sensitivity.
-// If s is not valid returns error.
+// convEndpoint converts s from a URL string to an endpoint if s is a valid
+// URL. Otherwise, "" and an error are returned.
+func convEndpoint(s string) (string, error) {
+	u, err := url.Parse(s)
+	if err != nil {
+		return "", err
+	}
+	return u.Host, nil
+}
+
+// convInsecure parses s as a URL string and returns if the connection should
+// use client transport security or not. If s is an invalid URL, false and an
+// error are returned.
 func convInsecure(s string) (bool, error) {
-	s = strings.ToLower(s)
-	if s != "true" && s != "false" {
-		return false, fmt.Errorf("can't convert %q to bool", s)
+	u, err := url.Parse(s)
+	if err != nil {
+		return false, err
 	}
+	return u.Scheme != "https", nil
+}
+
+func loadInsecureFromEnvEndpoint(envEndpoint []string) resolver[bool] {
+	return func(s setting[bool]) setting[bool] {
+		if s.Set {
+			// Passed, valid, options have precedence.
+			return s
+		}
 
-	return s == "true", nil
+		for _, key := range envEndpoint {
+			if vStr := os.Getenv(key); vStr != "" {
+				u, err := url.Parse(vStr)
+				if err != nil {
+					otel.Handle(fmt.Errorf("invalid %s value %s: %w", key, vStr, err))
+					continue
+				}
+
+				return insecureFromScheme(s, u.Scheme)
+			}
+		}
+		return s
+	}
 }
 
 // convHeaders converts the OTel environment variable header value s into a
@@ -522,33 +551,6 @@ func loadCertificates(certPath, keyPath string) ([]tls.Certificate, error) {
 	return []tls.Certificate{crt}, nil
 }
 
-// loadEnvEndpoint loads and returns endpoint from envEndpoint variables.
-// It parses env value with url.Parse.
-func loadEnvEndpoint(fn func(*url.URL)) resolver[string] {
-	return func(s setting[string]) setting[string] {
-		if s.Set {
-			// Passed, valid, options have precedence.
-			return s
-		}
-
-		for _, key := range envEndpoint {
-			if vStr := os.Getenv(key); vStr != "" {
-				u, err := url.Parse(vStr)
-				if err != nil {
-					otel.Handle(fmt.Errorf("invalid %s value %s: %w", key, vStr, err))
-					continue
-				}
-
-				s.Value = u.Host
-				s.Set = true
-				fn(u)
-				return s
-			}
-		}
-		return s
-	}
-}
-
 // insecureFromScheme return setting if the connection should
 // use client transport security or not.
 // Empty scheme doesn't force insecure setting.
```

Adding two small methods with not many repeating codes seems good to me.